src/array.jl

```diff
@@ -54,7 +58,10 @@ function unsafe_free!(a::CuArray)
     ptr = pointer(a)
     if isvalid(ptr.ctx)
         @trace("Finalizing CuArray object at $(Base.pointer_from_objref(a))")
-        Mem.free(ptr)
+        if Mem.refcount(ptr) == 1
```

Comment:
in addition, you can make `Mem.malloc` call `retain` such that you only need the additional `retain` call from the unsafe `CuArray` constructor.